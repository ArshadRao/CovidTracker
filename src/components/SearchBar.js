import React from 'react';
import {
  View,
  StyleSheet,
  TextInput,
  Image,
  Dimensions,
} from 'react-native';
import searchLogo from '../assets/search.png';
import logo from '../assets/logo.png';


const deviceSize = Dimensions.get('window');
const colors = {
  gray: 'gray',
  searchBackColor: '#eceff1',
  fontColor: 'black',
  tintColor: 'rgb(242, 63, 52)',
};

const styles = StyleSheet.create({
  iconImage: {
    margin: 5,
    width: deviceSize.width * 0.1,
    height: deviceSize.width * 0.1,
    tintColor: colors.tintColor,
  },
  container: {
    alignSelf: 'center',
    height: 30,
    flex: 1,
    margin: 5,
    paddingHorizontal: 5,
    borderRadius: 5,
    backgroundColor: colors.searchBackColor,
    flexDirection: 'row',
  },
  inputStyle: {
    flex: 1,
    padding: 0,
    color: colors.fontColor,
  },
  logo: {
    width: deviceSize.width * 0.04,
    height: deviceSize.height * 0.03,
    resizeMode: 'contain',
    tintColor: colors.gray,
  },
});

const SearchBar = (props) => (
  <View style={{ flexDirection: 'row' }}>
    <View style={{ flexDirection: 'row', alignItems: 'center' }}>
      <Image
        style={styles.iconImage}
        source={logo}
      />
    </View>

    <View style={styles.container}>
      <TextInput
        style={styles.inputStyle}
        placeholder={props.label}
        placeholderTextColor={colors.gray}
        onChangeText={props.onSearch}
      />
      <Image source={searchLogo} style={styles.logo} />
    </View>
  </View>
);

export default SearchBar;
