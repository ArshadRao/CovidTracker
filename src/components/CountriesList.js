import React, { Component } from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  Text,
  View,
  FlatList,
  Image,
  Dimensions,
  SafeAreaView,
} from 'react-native';
import axios from 'axios';
import ListItem from './ListItem';
import Header from './Header';
import SearchBar from './SearchBar';
import searchLogo from '../assets/search.png';


const deviceSize = Dimensions.get('window');
const separatorColor = '#eceff1';
const surfaceColor = 'white';
const colorsGray = 'gray';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  horizontal: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    padding: 10,
  },
  separator: {
    borderColor: separatorColor,
    borderWidth: 1,
  },
  mainContainer: {
    flex: 1,
    backgroundColor: surfaceColor,
  },
  noResultImage: {
    width: deviceSize.width * 0.1,
    height: deviceSize.height * 0.1,
    resizeMode: 'contain',
    tintColor: colorsGray,
  },
  noResultTitle: {
    color: colorsGray,
    textAlign: 'center',
  },
});

export function PureCountriesList() {
  const [loading, setLoading] = React.useState(true);
  const [dataList, setDataList] = React.useState([]);
  const [filteredData, setFilteredData] = React.useState([]);

  const fetchData = async () => {
    setLoading(true);
    const response = await axios.get('https://corona.lmao.ninja/v2/countries');
    setLoading(false);
    setDataList(response.data);
    setFilteredData(response.data);
  };

  React.useEffect(() => {
    fetchData();
  }, []);

  const renderCountries = ({ item }) => <ListItem item={item} />;

  const searchByName = (text) => {
    const filteredList = dataList.filter((item) => {
      const itemData = item.country.toUpperCase();
      const textData = text.toUpperCase();

      return itemData.indexOf(textData) > -1;
    });
    setFilteredData(filteredList);
  };

  const renderSeparator = () => <View style={styles.separator} />;

  const renderEmpty = (
    <View style={styles.center}>
      <Image source={searchLogo} style={styles.noResultImage} />
      <Text style={styles.noResultTitle}>List is Empty</Text>
    </View>
  );

  if (loading) {
    return (
      <View style={[styles.container, styles.horizontal]}>
        <ActivityIndicator size="large" color="#0000ff" />
      </View>
    );
  }

  return (
    <SafeAreaView style={styles.mainContainer}>
      <Header title="Countries" />
      <SearchBar label="Search.." onSearch={searchByName} />
      <FlatList
        data={filteredData}
        refreshing={loading}
        initialNumToRender={10}
        onRefresh={fetchData}
        renderItem={renderCountries}
        ListEmptyComponent={renderEmpty}
        contentContainerStyle={{ flexGrow: 1 }}
        ItemSeparatorComponent={renderSeparator}
        keyExtractor={(_, index) => index.toString()}
      />
    </SafeAreaView>
  );
}

export default PureCountriesList;
