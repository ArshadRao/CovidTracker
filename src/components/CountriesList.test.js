import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';
import { PureCountriesList as CountriesList } from './CountriesList';
import { FlatList } from 'react-native';
import SearchBar from './SearchBar';

const mockUseEffectOnce = () => {
  jest.spyOn(React, 'useEffect').mockImplementationOnce((fn) => fn());
};

const mockUseStateOnce = (state, setState = jest.fn()) => {
  jest.spyOn(React, 'useState').mockImplementationOnce(() => [state, setState]);
};

const data = [
  {
    updated: 1587140875474,
    country: 'Italy',
    cases: 168941,
    todayCases: 3786,
    deaths: 22170,
    todayDeaths: 525,
    recovered: 40164,
    active: 106607,
    critical: 2936,
    tests: 1178403,
    population: 60390869,
    continent: 'Europe',
  },
  {
    updated: 1618868527825,
    country: 'Brazil',
    cases: 13973695,
    todayCases: 30624,
    deaths: 374682,
    todayDeaths: 1240,
    recovered: 12391599,
    todayRecovered: 0,
    active: 1207414,
    critical: 8318,
    tests: 28600000,
    population: 213761802,
    continent: 'South America',
  },
];

jest.mock('axios', () => ({
  get: async () => ({ data }),
}));

describe('CountriesList component', () => {
  it('should render component properly', () => {
    const wrapper = shallow(<CountriesList />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with Flatlist', async () => {
    await mockUseEffectOnce();
    const setLoading = jest.fn();
    mockUseStateOnce(false, setLoading);
    mockUseStateOnce(data);
    mockUseStateOnce(data);
    const wrapper = shallow(<CountriesList />);
    expect(wrapper).toMatchSnapshot();
    expect(setLoading).toHaveBeenCalled();
  });

  it('should render listItem', () => {
    mockUseEffectOnce();
    mockUseStateOnce(false);
    mockUseStateOnce(data);
    mockUseStateOnce(data);
    const wrapper = shallow(<CountriesList />);
    const listItem = wrapper.find(FlatList).props().renderItem({ item: data[0] });
    expect(React.isValidElement(listItem)).toBe(true);
  });

  it('should call keyExtractor', () => {
    mockUseEffectOnce();
    mockUseStateOnce(false);
    mockUseStateOnce(data);
    mockUseStateOnce(data);
    const wrapper = shallow(<CountriesList />);
    const index = wrapper.find(FlatList).props().keyExtractor('', 0);
    expect(index).toBe('0');
  });

  it('should filter list on search', () => {
    const setFilteredData = jest.fn();
    mockUseEffectOnce();
    mockUseStateOnce(false);
    mockUseStateOnce(data);
    mockUseStateOnce(data, setFilteredData);
    const wrapper = shallow(<CountriesList />);
    wrapper.find(SearchBar).props().onSearch('Brazil');
    expect(setFilteredData).toHaveBeenCalledWith([data[1]]);
  });
});
