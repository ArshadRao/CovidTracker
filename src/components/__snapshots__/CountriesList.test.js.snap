// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountriesList component should render component properly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "padding": 10,
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#0000ff"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`CountriesList component should render component with Flatlist 1`] = `
<ForwardRef(SafeAreaView)
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <Header
    title="Countries"
  />
  <SearchBar
    label="Search.."
    onSearch={[Function]}
  />
  <FlatList
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={
      <View>
        <Image
          source={
            Object {
              "process": [Function],
            }
          }
          style={
            Object {
              "height": 133.4,
              "resizeMode": "contain",
              "tintColor": "gray",
              "width": 75,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "gray",
              "textAlign": "center",
            }
          }
        >
          List is Empty
        </Text>
      </View>
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={
      Array [
        Object {
          "active": 106607,
          "cases": 168941,
          "continent": "Europe",
          "country": "Italy",
          "critical": 2936,
          "deaths": 22170,
          "population": 60390869,
          "recovered": 40164,
          "tests": 1178403,
          "todayCases": 3786,
          "todayDeaths": 525,
          "updated": 1587140875474,
        },
        Object {
          "active": 1207414,
          "cases": 13973695,
          "continent": "South America",
          "country": "Brazil",
          "critical": 8318,
          "deaths": 374682,
          "population": 213761802,
          "recovered": 12391599,
          "tests": 28600000,
          "todayCases": 30624,
          "todayDeaths": 1240,
          "todayRecovered": 0,
          "updated": 1618868527825,
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    onRefresh={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</ForwardRef(SafeAreaView)>
`;
